File: src/main/java/io/vproxy/vfx/animation/AnimationNode.java
package io.vproxy.vfx.animation;

import io.vproxy.vfx.util.algebradata.AlgebraData;
import io.vproxy.vfx.util.graph.GraphNode;

public class AnimationNode<T extends AlgebraData<T>> extends GraphNode<AnimationNode<T>> {
    public final String name;
    public final T value;
    public final AnimationStateTransferFinishCallback<T> stateTransferFinish;

    public AnimationNode(String name, T value) {
        this(name, value, (from, to) -> {
        });
    }

    public AnimationNode(String name, T value,
                         AnimationStateTransferFinishCallback<T> stateTransferFinish) {
        super(name);
        this.name = name;
        this.value = value;
        this.stateTransferFinish = stateTransferFinish;
    }
}

File: src-ts/src/animation/AnimationNode.ts
import type { AlgebraData } from '../util/algebradata/AlgebraData.js';
import { GraphNode } from '../commons-graph/Graph.js';
import type { AnimationStateTransferFinishCallback } from './AnimationStateTransferFinishCallback.js';

const noopFinish: <T extends AlgebraData<T>>(from: AnimationNode<T>, to: AnimationNode<T>) => void =
  () => { /* no-op */ };

export class AnimationNode<T extends AlgebraData<T>> extends GraphNode<AnimationNode<T>> {
  override readonly name: string;
  readonly value: T;
  readonly stateTransferFinish: AnimationStateTransferFinishCallback<T>;

  constructor(name: string, value: T);
  constructor(name: string, value: T, stateTransferFinish: AnimationStateTransferFinishCallback<T>);
  constructor(
    name: string,
    value: T,
    stateTransferFinish: AnimationStateTransferFinishCallback<T> = noopFinish as AnimationStateTransferFinishCallback<T>,
  ) {
    super(name);
    this.name = name;
    this.value = value;
    this.stateTransferFinish = stateTransferFinish;
  }
}
